lib/ood_core/job/adapters/linux_host/launcher.rb

```diff
@@ -0,0 +1,262 @@
+require 'erb'
+require 'etc'
+require 'pathname'
+require 'securerandom'
+require 'shellwords'
+require 'time'
+
+# Object used for simplified communication SSH hosts
+#
+# @api private
+class OodCore::Job::Adapters::LinuxHost::Launcher
+  attr_reader :contain, :debug, :site_timeout, :session_name_label, :singularity_bin,
+    :site_singularity_bindpath, :default_singularity_image, :ssh_hosts,
+    :strict_host_checking, :submit_host, :tmux_bin, :username
+  # The root exception class that all LinuxHost adapter-specific exceptions inherit
+  # from
+  class Error < StandardError; end
+
+  UNIT_SEPARATOR = "\x1F"
+
+  # @param debug Whether the adapter should be used in debug mode
+  # @param site_timeout [#to_i] A period after which the job should be killed or nil
+  # @param singularity_bin  Path to the Singularity executable
+  # @param singularity_bindpath A comma delimited string of host paths to bindmount into the guest; sets SINGULARITY_BINDPATH environment variable
+  # @param singularity_image [#to_s] Path to the Singularity image
+  # @param ssh_hosts List of hosts to check when scanning for running jobs
+  # @param strict_host_checking Allow SSH to perform strict host checking
+  # @param submit_host The SSH-able host
+  # @param tmux_bin [#to_s] Path to the tmux executable
+  def initialize(
+    contain: false,
+    debug: false,
+    site_timeout: nil,
+    singularity_bin:,
+    singularity_bindpath: '/etc,/media,/mnt,/opt,/run,/srv,/usr,/var,/users',
+    singularity_image:,
+    ssh_hosts:,
+    strict_host_checking: false,
+    submit_host:,
+    tmux_bin:,
+    **_
+  )
+    @contain = !! contain
+    @debug = !! debug
+    @site_timeout = site_timeout.to_i
+    @session_name_label = 'launched-by-ondemand'
+    @singularity_bin = Pathname.new(singularity_bin)
+    @site_singularity_bindpath = singularity_bindpath.to_s
+    @default_singularity_image = Pathname.new(singularity_image)
+    @ssh_hosts = ssh_hosts
+    @strict_host_checking = strict_host_checking
+    @submit_host = submit_host
+    @tmux_bin = tmux_bin
+    @username = Etc.getlogin
+  end
+
+  # @param hostname [#to_s] The hostname to submit the work to
+  # @param script [OodCore::Job::Script] The script object defining the work
+  def start_remote_session(script)
+    cmd = ssh_cmd(submit_host)
+
+    session_name = unique_session_name
+    output = call(*cmd, stdin: wrapped_script(script, session_name))
+    hostname = output.strip
+
+    "#{session_name}@#{hostname}"
+  end
+
+  def stop_remote_session(session_name, hostname)
+    cmd = ssh_cmd(hostname)
+
+    # discover tmux sessions
+    # discover the PID for the given session
+    # get all children of that PID
+    # find the sinit process for the tmux PID
+    # kill the sinit process
+    kill_cmd = <<~SCRIPT
+    kill $(pstree -p $(tmux list-panes -aF '#\{session_name} \#{pane_pid}' | grep '#{session_name}' | cut -f 2 -d ' ') | grep -o 'sinit([[:digit:]]*' | grep -o '[[:digit:]]*')
```

Comment:
There is a lot going on in one line. Just for readability maybe split each `$()` execution into separate variable:

```
PAIN_PIDS=$(tmux list-panes -aF '#\{session_name} \#{pane_pid}' | grep '#{session_name}' | cut -f 2 -d ' ')
PIDS=$(pstree -p "${PAIN_PIDS}" | grep -o 'sinit([[:digit:]]*' | grep -o '[[:digit:]]*')
kill "${PIDS}"
```
Might have to adjust double quotes around variables depending on what output looks like.

What do session names look like? Would it be necessary for `grep '#{session_name}'` to be `grep -e '^#{session_name}' just to ensure correct match?

Also is the output of each `$()` going to contain new lines? Will the commands they are passed to handle newlines? Like will you end up saying this?

```
kill 123
124
125
```

Reply:
Session names are purely programmatic, and look like `launched_by_ondemand-$UUID`. The tmux pane format strings look a lot like Ruby string interpolation which is a little confusing hence the escapes in one `session_name` but not the second. Since the session name is interpolated into the heredoc by Ruby there should only ever be one match; so `kill` would only get a single PID.

Reply:
What about the output of `pstree`? WIll that contain newlines or just one PID? I would assume more than one PID is output by `pstree` or does the greps after ensure just one PID is returned?

Reply:
By looking for `sinit(` I'm able to cut it down to a single PID.